File: 965-univalued-binary-tree/965-univalued-binary-tree.js
// Approach 1: Depth-First Search

// Time Complexity: O(N) 
// N is the number of nodes in the given tree.

// Space Complexity: O(N)

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isUnivalTree = function(root) {
    var dfs = function(node, val) {
        if (node === null) {
            return true;
        }
        if (node.val !== val) {
            return false;
        }
        return dfs(node.left, val) && dfs(node.right, val)
    }
    return dfs(root, root.val)
};

File: 965-univalued-binary-tree/965-univalued-binary-tree.py
# Approach 2: Recursion

# A tree is univalued if both its children are univalued, plus the root node has the same value as the child nodes

# Time Complexity: O(N) 
# N is the number of nodes in the given tree.

# Space Complexity: O(H)
# H is the height of the given tree.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isUnivalTree(self, root: Optional[TreeNode]) -> bool:
        left_correct = not root.left or (root.val == root.left.val and self.isUnivalTree(root.left))
        right_correct = not root.right or (root.val == root.right.val and self.isUnivalTree(root.right))
        return left_correct and  right_correct
        